# Dataset Tutorial
This notebook gives a short tutorial on how to use the pytorch dataset I implemented.
You find the code in `src/data/dataset.py` in the `CnvDataset` class. 

Let's first start by importing some packages we might need

In [1]:
import pandas as pd
from pathlib import Path

# add this to you notebook so it automatically reloads code you changed in a
# python file after importing this code
%load_ext autoreload
%autoreload 2

Now we also need to import the `CnvDataset` class.
Since the path (relative to the git repository root) to this notebook is `preprocessing/dataset_example.ipynb`, we need to add the parent directory to our system path in order to import software from there.
Think of it like this: We need to tell the notebook the relative path to the software folder `src` in order to import software from there.

In [2]:
import sys
sys.path.append('..') # add the parent directory to system path
from src.data.dataset import CnvDataset

Great :)
Now that this is out of the way, let's define some important paths for files we need to work with.

In [3]:
# directories we will need
git_root = Path('..')
data_root = git_root / 'data'
assert data_root.exists()

# dataset split files
b1_train_path = data_root / 'splits' / 'batch1_training_filtered.tsv'
b1_val_path = data_root / 'splits' / 'batch1_val_filtered.tsv'
b1_test_path = data_root / 'splits' / 'batch1_test_filtered.tsv'

Alright, now we are almost ready to use the `CnvDataset`.
One last thing thats is missing, is the path to the directory that stores the embedding files for the dataset we want to use.
All dataset paths follow the same pattern:
```
data/embeddings/batch_<batch_number>/<dataset_type>/<embedding_mode>
```
where:
* `batch_<batch_number>` is either `batch_1` or `batch_2`
* `<dataset_type>` is one of `train`, `val` or `test`
* `<embedding_mode>` is one of `single_gene_barcode`, `gene_concat` or `barcode_channel`

Please note, that the `<embedding_mode>` will be added automatically.
You don't need to add it to the dataset path, just change the `embedding_mode` parameter for the `CnvDataset` class.
Also, please make sure that the directory actually exists.
However, the python code will raise an exception if does not find any embedding files.

OK. Now let's define the dataset we want to use.
In this example I chose the validation set of batch 1.

In [4]:
dataset_root = data_root / 'embeddings' / 'batch_1' / 'val'

Next, we read the validation split data frame using pandas.

In [5]:
b1_val_df = pd.read_csv(b1_val_path, sep='\t')

In [6]:
b1_val_dataset = CnvDataset(
    root=dataset_root,
    data_df=b1_val_df,
    embedding_mode='single_gene_barcode'
)

Using 51 barcodes
Using 1093 genes
No embedding files for 932 data points in ../data/embeddings/batch_1/val/single_gene_barcode!
Found 38082 embedding files with no target value in ../data/embeddings/batch_1/val/single_gene_barcode!


Your output should look something like:
```
Using 51 barcodes
Using 1093 genes
No embedding files for 932 data points in ../data/embeddings/batch_1/val/single_gene_barcode!
Found 38082 embedding files with no target value in ../data/embeddings/batch_1/val/single_gene_barcode!
```

This means that from the datapoint with target values in the dataset, we are missing 932 embedding files.
On the other hand, there are 38082 embedding files with no target value associated.

Now you should be able to get the number of data points and the first rows of the data frame by using the string representaiton of the dataset variable.

In [34]:
print(str(b1_val_dataset))

<class 'src.data.dataset.CnvDataset'> with 8022 datapoints
               barcode          gene_id  expression_count classification  \
15  AAAGGTTAGGGTGGAT-1  ENSG00000020577          0.407756            low   
17  AAAGGTTAGGGTGGAT-1  ENSG00000021645          2.146118           high   
26  AAAGGTTAGGGTGGAT-1  ENSG00000030582          0.407756            low   
27  AAAGGTTAGGGTGGAT-1  ENSG00000033327          0.696581            low   
32  AAAGGTTAGGGTGGAT-1  ENSG00000038427          1.103188           high   

                                       embedding_path  
15  ../data/embeddings/batch_1/val/single_gene_bar...  
17  ../data/embeddings/batch_1/val/single_gene_bar...  
26  ../data/embeddings/batch_1/val/single_gene_bar...  
27  ../data/embeddings/batch_1/val/single_gene_bar...  
32  ../data/embeddings/batch_1/val/single_gene_bar...  


Also we should be able to get the embedding and the classification label from the dataset using an index (just like a list).

In [35]:
b1_val_dataset[0]

{'embedding': tensor([[0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 1,  ..., 0, 1, 0],
         ...,
         [0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0]], dtype=torch.uint8),
 'label': 'low'}

To use the dataset for regression you need to set the `target_type` parameter for the `CnvDataset` class to `'regression'`.

In [7]:
b1_val_dataset = CnvDataset(
    root=dataset_root,
    data_df=b1_val_df,
    target_type='regression'
)

Using 51 barcodes
Using 1093 genes
No embedding files for 932 data points in ../data/embeddings/batch_1/val/single_gene_barcode!
Found 38082 embedding files with no target value in ../data/embeddings/batch_1/val/single_gene_barcode!


In [8]:
b1_val_dataset[0]

{'embedding': tensor([[0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 1,  ..., 0, 1, 0],
         ...,
         [0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0]], dtype=torch.uint8),
 'target': np.float64(0.40775576)}

Now let's load the training set for a regression use case.

In [13]:
train_set_root = data_root / 'embeddings' / 'batch_1' / 'training'
b1_train_df = pd.read_csv(b1_train_path, sep='\t')
b1_train_dataset = CnvDataset(
    root=train_set_root,
    data_df=b1_train_df,
    target_type='regression'
)

Using 356 barcodes
Using 1595 genes
No embedding files for 6207 data points in ../data/embeddings/batch_1/training/single_gene_barcode!
Found 386170 embedding files with no target value in ../data/embeddings/batch_1/training/single_gene_barcode!


In [14]:
b1_train_dataset

<class 'src.data.dataset.CnvDataset'> with 53134 datapoints

And here a example using the test dataset for classification.

In [10]:
test_set_root = data_root / 'embeddings' / 'batch_1' / 'test'
b1_test_df = pd.read_csv(b1_test_path, sep='\t')
b1_test_dataset = CnvDataset(
    root=test_set_root,
    data_df=b1_test_df,
    target_type='regression'
)

Using 102 barcodes
Using 1235 genes
No embedding files for 1966 data points in ../data/embeddings/batch_1/test/single_gene_barcode!
Found 87575 embedding files with no target value in ../data/embeddings/batch_1/test/single_gene_barcode!


In [15]:
b1_test_dataset

<class 'src.data.dataset.CnvDataset'> with 16669 datapoints

## Computing Embeddings from scratch
This section covers (re-)computing embeddings using the `CnvDataset`.
For this we need a little bit more information than before.
Like previouly, let's start by defining some paths to relevant files and directories. 

In [9]:
# directories we will need
out_root = git_root / 'out'

# files we will need
genome_fasta = data_root / 'reference' / 'GRCh38.d1.vd1.fa'
assert genome_fasta.exists()
overlap_path = data_root / 'overlap_genes_peaks.tsv'
assert overlap_path.exists()
epiAneufinder_path = out_root / 'epiAneufinder' / 'epiAneuFinder_results.tsv'
assert epiAneufinder_path.exists()

In [7]:
b1_df = pd.read_csv(b1_val_path, sep='\t')

In [ ]:
# compute all embeddings for batch 1
b1_dataset = CnvDataset(
    root=data_root / 'embeddings' / 'batch_1' / 'val' ,
    data_df=b1_df,
    fasta_path=genome_fasta,
    atac_path=overlap_path,
    cnv_path=epiAneufinder_path,
    embedding_mode='single_gene_barcode',
    force_recompute=True
)

In [ ]:
b1_dataset[0]

In [ ]:
b1_df

barcode          gene_id  expression_count classification  \
0      AAACATGCAGGATGGC-1  ENSG00000084070          0.371921            low   
1      AAACATGCAGGATGGC-1  ENSG00000127124          0.642400            low   
2      AAACATGCAGGATGGC-1  ENSG00000269113          1.179453           high   
3      AAACATGCAGGATGGC-1  ENSG00000173406          0.371921            low   
4      AAACATGCAGGATGGC-1  ENSG00000226476          0.371921            low   
...                   ...              ...               ...            ...   
49311  TTTGTTGGTACCAGGT-1  ENSG00000198899          2.295555           high   
49312  TTTGTTGGTACCAGGT-1  ENSG00000198938          1.884219           high   
49313  TTTGTTGGTACCAGGT-1  ENSG00000198840          0.749642            low   
49314  TTTGTTGGTACCAGGT-1  ENSG00000198886          1.698359           high   
49315  TTTGTTGGTACCAGGT-1  ENSG00000198727          1.173251           high   

                                          embedding_path  
0      ../data/embeddings/batch_1/val/single_gene_bar...  
1      ../data/embeddings/batch_1/val/single_gene_bar...  
2      ../data/embeddings/batch_1/val/single_gene_bar...  
3      ../data/embeddings/batch_1/val/single_gene_bar...  
4      ../data/embeddings/batch_1/val/single_gene_bar...  
...                                                  ...  
49311  ../data/embeddings/batch_1/val/single_gene_bar...  
49312  ../data/embeddings/batch_1/val/single_gene_bar...  
49313  ../data/embeddings/batch_1/val/single_gene_bar...  
49314  ../data/embeddings/batch_1/val/single_gene_bar...  
49315  ../data/embeddings/batch_1/val/single_gene_bar...  

[49316 rows x 5 columns]

In [ ]:
b1_dataset.data_df

Empty DataFrame
Columns: [barcode, gene_id, expression_count, classification, embedding_path]
Index: []

In [ ]:
[d for d in b1_dataset.root_path.iterdir()][:10]

[PosixPath('../data/embeddings/batch_1/val/single_gene_barcode/AGCCGCTAGAATGACG-1'),
 PosixPath('../data/embeddings/batch_1/val/single_gene_barcode/CATTATCTCGCGACAC-1'),
 PosixPath('../data/embeddings/batch_1/val/single_gene_barcode/CATGAGGCACGGTTTA-1'),
 PosixPath('../data/embeddings/batch_1/val/single_gene_barcode/GCTAAGTTCGGGCCAT-1'),
 PosixPath('../data/embeddings/batch_1/val/single_gene_barcode/AGTTGCGTCGATATTG-1'),
 PosixPath('../data/embeddings/batch_1/val/single_gene_barcode/AATTACCCAGCAAGTG-1'),
 PosixPath('../data/embeddings/batch_1/val/single_gene_barcode/GGACCGAAGCGATAAG-1'),
 PosixPath('../data/embeddings/batch_1/val/single_gene_barcode/ACCTACCTCGGCTAGC-1'),
 PosixPath('../data/embeddings/batch_1/val/single_gene_barcode/TGATGAACAAGGCCAA-1'),
 PosixPath('../data/embeddings/batch_1/val/single_gene_barcode/TGAACAGAGCACTTGG-1')]

In [ ]:
print(b1_dataset)

<class 'src.data.dataset.CnvDataset'> with 0 datapoints


In [ ]:
b1_df.iloc[0]['embedding_path']

PosixPath('../data/embeddings/batch_1/val/single_gene_barcode/AAACATGCAGGATGGC-1/ENSG00000084070.mtx')